nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp

```diff
@@ -72,6 +88,9 @@ class MotionModel
     const bool is_holo = isHolonomic();
     for (unsigned int i = 0; i != state.vx.shape(0); i++) {
       for (unsigned int j = 1; j != state.vx.shape(1); j++) {
+        float max_feasible_vx = state.vx(i, j-1) + model_dt_*accl_constraints_.ax_max;
```

Comment:
I'm curious what the runtime change of this action is. Can you benchmark the performance changes after adding in the wz / vy acceleration constraints

Reply:
I benchmarked the performance changes by running the simulation using default tb3_simulation_launch.py.
 
MPPI parameters:
`time_steps: 56
 model_dt: 0.05
 batch_size: 2000
 ax_max: 3.5
 ax_min: -3.0
 ay_max: 0.0
 ay_min: -0.0
 thetaz_max: 3.0
 thetaz_min: -3.0`

**Without acceleration constraints:**
`[component_container_isolated-5] Total duration in ms = 10931
[component_container_isolated-5] No. of loops = 927
[component_container_isolated-5] Average time in ms = 11.7918
`
**With acceleration constraints:**
`[component_container_isolated-5] Total duration in ms = 12858
[component_container_isolated-5] No. of loops = 1001
[component_container_isolated-5] Average time in ms = 12.8452
`
The increment in the execution time of the control loop was around 12.84 - 11.79  = 1.05 ms averaging over 1000 loops.

Reply:
Benchmark results:

Running ./optimizer_benchmark
Run on (12 X 4100 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x6)
  L1 Instruction 32 KiB (x6)
  L2 Unified 256 KiB (x6)
  L3 Unified 9216 KiB (x1)
Load Average: 0.77, 1.97, 1.46

Iterations: 2
Batch size: 1000
Time steps: 50
Path points: 150

** With acceleration constraints for all axes**
---------------------------------------------------------------------
Benchmark                           Time             CPU   Iterations
---------------------------------------------------------------------
BM_DiffDrivePointFootprint   4.87 ms         4.87 ms          115
BM_DiffDrive                         5.03 ms         5.03 ms          123
BM_Omni                              5.29 ms         5.29 ms          126
BM_Ackermann                    4.91 ms         4.91 ms          130

** Without acceleration constraints **
---------------------------------------------------------------------
Benchmark                           Time             CPU   Iterations
---------------------------------------------------------------------
BM_DiffDrivePointFootprint   4.28 ms         4.28 ms          149
BM_DiffDrive                         4.58 ms         4.58 ms          142
BM_Omni                              4.79 ms         4.79 ms          118
BM_Ackermann                    5.59 ms         5.59 ms          100

Reply:
Can you average a 3-5 runs together to get an average value? I want to get a good sense of this. But overall seems not terrible. 

Please also:
- Precompute all the `model_dt_ * accel_constraints_.ax_max`'s (and similar) before the loop and use those variables. Given the size of the tensor, removing those multiplications from the loop can speed things up
- `state.wz(i, j - 1)` when you access things, store them as references (`float & wz_curr = state.wz(i, j - 1)`) and use those. When you do those `state.wz(i, j - 1)`, that's a lookup that adds compute time. Look up each once and use the stored value reference from then on.

Reply:
yes sure. I'll modify the code to avoid lookups. Just a quick question, why we aren't publishing the generated local trajectory(from final control sequences) to visualize it in RViz?

Reply:
We are! Its done in another file though :-) Check out [the ](https://github.com/ros-navigation/navigation2/blob/main/nav2_mppi_controller/src/trajectory_visualizer.cpp)

Reply:
Without acceleration constraints:
| Benchmark | run 1 | run 2 | run 3 | run 4 | run 5 |  Average |
| --------------- | ------- | ------- | ------- | ------- | ------- | ------------ |
| BM_DiffDrivePointFootprint  | 4.75 ms | 4.21 ms | 4.23 ms | 4.84 ms | 4.31 ms  | 4.468 ms |
| BM_DiffDrive | 4.72 ms | 4.67 ms | 4.67 ms | 4.87 ms | 4.61 ms | 4.708 ms |
| BM_Omni  | 5.47 ms | 4.89 ms | 4.98 ms | 5.45 ms | 5.02 ms | 5.162 ms |
| BM_Ackermann | 4.45 ms | 5.35 ms | 5.34 ms | 4.52 ms | 5.37 ms | 5.005 ms

With acceleration constraints for all axes:
| Benchmark | run 1 | run 2 | run 3 | run 4 | run 5 |  Average | %change|
| --------------- | ------- | ------- | ------- | ------- | ------- | ------------ | --------------- |
| BM_DiffDrivePointFootprint  | 4.81 ms | 4.94 ms | 4.86 ms | 4.81 ms | 4.92 ms  | 4.868 ms | 8.95% |
| BM_DiffDrive | 5.00 ms | 5.07 ms | 5.05 ms | 5.10 ms | 5.09 ms | 5.062 ms | 7.51% |
| BM_Omni  | 5.36 ms | 5.24 ms | 5.37 ms | 5.43 ms | 5.24 ms | 5.328 ms | 3.21% |
| BM_Ackermann | 4.90 ms | 4.94 ms | 4.94 ms | 5.02 ms | 4.90 ms | 4.940 ms | -1.29% |

Reply:
@SteveMacenski any comments/suggestions on the benchmark results? I've also tried this to check the performance `for(unsigned int j = 1; j != state.vx.shape(1); j++) {
      auto vx_curr = xt::view(state.vx, xt::all(), j - 1);
      auto cvx_curr = xt::view(state.cvx, xt::all(), j - 1);
      xt::noalias(cvx_curr) = xt::clip(cvx_curr, vx_curr + min_delta_vx, vx_curr + max_delta_vx);
      xt::noalias(xt::view(state.vx, xt::all(), j)) = cvx_curr;

      auto wz_curr = xt::view(state.wz, xt::all(), j - 1);
      auto cwz_curr = xt::view(state.cwz, xt::all(), j - 1);
      xt::noalias(cwz_curr) = xt::clip(cwz_curr, wz_curr - max_delta_wz, wz_curr + max_delta_wz);
      xt::noalias(xt::view(state.wz, xt::all(), j)) = cwz_curr;

      if(is_holo) {
        auto vy_curr = xt::view(state.vy, xt::all(), j - 1);
        auto cvy_curr = xt::view(state.cvy, xt::all(), j - 1);
        xt::noalias(cvy_curr) = xt::clip(cvy_curr, vy_curr + min_delta_vy, vy_curr + max_delta_vy);
        xt::noalias(xt::view(state.vy, xt::all(), j)) = cvy_curr;
      }  
    }`
but it was much slower. If we somehow parallelize the computation for rows(1000-2000 rows), could that reduce the compute time significantly?

Reply:
See https://github.com/ros-navigation/navigation2/pull/4352#discussion_r1612242099, I think that should get us back down to sub ~1% increase hopefully

Reply:
benchmarks with a batch size of 2000, as results were more stable compared to a batch size of 1000.

Without acceleration constraints:
| Benchmark |  Average |
| --------------- | ---------- |
| BM_DiffDrivePointFootprint  | 11.22 ms |
| BM_DiffDrive |  12.06 ms |
| BM_Omni  | 14.66 ms |
| BM_Ackermann | 12.54 ms |

With acceleration constraints for all axes:
| Benchmark | Average |
| --------------- | ----------- |
| BM_DiffDrivePointFootprint  | 12.8 ms |
| BM_DiffDrive | 11.92 ms |
| BM_Omni  | 14.1 ms |
| BM_Ackermann |  12.36 ms |

After reducing all the redundant lookups, surprisingly compute time with acceleration constraints becomes lesser than the compute time without acceleration constraints. (Except for the 1st row, DiffDrive with Footprint).

Reply:
I think those are within the range of measurement noise - this is perfectly good, thanks for working with me to get this compute minimized :-)